File: ios/Classes/scan/ScanUtils.m

#import "ScanView.h"
#import "ScanUtils.h"

@implementation ScanUtils
+ (void)scanImagePath:(FlutterMethodCall*)call result:(FlutterResult)result{
    NSString * path=[call.arguments valueForKey:@"path"];
    if([path isKindOfClass:[NSNull class]]){
        result(nil);
        return;
    }
    //加载文件
    NSFileHandle * fh=[NSFileHandle fileHandleForReadingAtPath:path];
    NSData * data=[fh readDataToEndOfFile];
    result([self getCode:data]);
}

+ (void)scanImageUrl:(FlutterMethodCall*)call result:(FlutterResult)result{
    NSString * url = [call.arguments valueForKey:@"url"];
    NSURL* nsUrl=[NSURL URLWithString:url];
    NSData * data=[NSURLConnection sendSynchronousRequest:[NSURLRequest requestWithURL:nsUrl] returningResponse:nil error:nil];
    result([self getCode:data]);
}

+ (void)scanImageMemory:(FlutterMethodCall*)call result:(FlutterResult)result{
    FlutterStandardTypedData * uint8list=[call.arguments valueForKey:@"uint8list"];
    result([self getCode:uint8list.data]);
}

+(NSDictionary *) getCode:(NSData *)data{
    if (data) {
        CIImage * detectImage=[CIImage imageWithData:data];
        CIDetector*detector = [CIDetector detectorOfType:CIDetectorTypeQRCode context:nil options:@{CIDetectorAccuracy: CIDetectorAccuracyHigh}];
        NSArray* feature = [detector featuresInImage:detectImage options: nil];
        if(feature.count==0){
            return nil;
        }else{
            for(int index=0;index<[feature count];index ++){
                CIQRCodeFeature * qrCode=[feature objectAtIndex:index];
                NSString *resultStr=qrCode.messageString;
                if(resultStr!=nil){
                    NSMutableDictionary *dict =[NSMutableDictionary dictionary];
                    [dict setValue:resultStr forKey:@"code"];
                    [dict setValue:AVMetadataObjectTypeQRCode forKey:@"type"];
                    return dict;
                }
                
            }
        }
    }
    return nil;
}

+(NSDictionary*) scanDataToMap:(AVMetadataMachineReadableCodeObject*) data{
    if (data == nil) {
        return nil;
    }
    NSMutableDictionary * result =[NSMutableDictionary dictionary];
    [result setValue:data.stringValue forKey:@"code"];
     [result setValue:data.type forKey:@"type"];
    return result;
}

@end

File: ios/Classes/scan/ScanUtils.swift
import Foundation
import AVFoundation
import Flutter
class ScanUtils {
    //本地图片识别
    public func scanImagePath(call: FlutterMethodCall, result: FlutterResult) {
        let path = call.arguments("path")
        if (path is NSNull) {
            result(nil)
            return
        }
        //加载文件
        let fh = FileHandle(forReadingAtPath: path ?? "")
        let data = fh?.readDataToEndOfFile()
        result(self.getCode(data))
    }
    //图片urls识别
    public  func scanImageUrl(call: FlutterMethodCall, result: FlutterResult) {
        let url = call.arguments(forKey: "url")
        let nsUrl = URL(string: url ?? "")
        var data: Data? = nil
        do {
            if let nsUrl = nsUrl {
                data = try NSURLConnection.sendSynchronousRequest(URLRequest(url: nsUrl), returning: nil)
            }
        } catch {
        }
        result(self.getCode(data))
    }
    //内存图片识别
    public func scanImageMemory(call: FlutterMethodCall, result: FlutterResult) {
        let uint8list = call.arguments.value(forKey: "uint8list") as? FlutterStandardTypedData
        result(self.getCode(uint8list?.data))
    }
    //获取二维码数据
    public  func getCode(data: Data) -> [AnyHashable : Any] {
        if data != nil {
            var detectImage: CIImage? = nil
            if let data = data {
                detectImage = CIImage(data: data)
            }
            let detector = CIDetector(ofType: CIDetectorTypeQRCode, context: nil, options: [
                CIDetectorAccuracy: CIDetectorAccuracyHigh
            ])
            var feature: [CIFeature]? = nil
            if let detectImage = detectImage {
                feature = detector?.features(in: detectImage, options: nil)
            }
            if feature?.count == 0 {
                return nil
            } else {
                for index in 0..<(feature?.count ?? 0) {
                    let qrCode = feature?[index] as? CIQRCodeFeature
                    let resultStr = qrCode?.messageString
                    if resultStr != nil {
                        var dict: [AnyHashable : Any] = [:]
                        dict["code"] = resultStr
                        dict["type"] = AVMetadataObject.ObjectType.qr
                        return dict
                    }
                }
            }
        }
        return nil
    }
    
    //二维码数据转换
    public func scanDataToMap(data: AVMetadataMachineReadableCodeObject) -> [AnyHashable : Any] {
        if data == nil {
            return nil
        }
        var result: [AnyHashable : Any] = [:]
        result["code"] = data?.stringValue
        result["type"] = data?.type
        return result
    }
    
}
